src/rpc/methods/f3.rs

```diff
@@ -0,0 +1,387 @@
+        (f3_tsk,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let tsk = f3_tsk.try_into()?;
+        let ts = ctx.chain_index().load_required_tipset(&tsk)?;
+        let parent = ctx.chain_index().load_required_tipset(ts.parents())?;
+        Ok(parent.into())
+    }
+}
+
+pub enum GetPowerTable {}
+impl RpcMethod<1> for GetPowerTable {
+    const NAME: &'static str = "F3.GetPowerTable";
+    const PARAM_NAMES: [&'static str; 1] = ["tipset_key"];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (F3TipSetKey,);
+    type Ok = Vec<F3PowerEntry>;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (f3_tsk,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        macro_rules! handle_miner_state_v12_on {
+            ($version:tt, $id_power_worker_mappings:ident, $ts:expr, $state:expr, $policy:expr) => {
+                fn map_err<E: Display>(e: E) -> fil_actors_shared::$version::ActorError {
+                    fil_actors_shared::$version::ActorError::unspecified(e.to_string())
+                }
+
+                let claims = $state.load_claims(ctx.store())?;
+                claims.for_each(|miner, claim| {
+                    if !claim.quality_adj_power.is_positive() {
+                        return Ok(());
+                    }
+
+                    let id = miner.id().map_err(map_err)?;
+                    let (_, ok) = $state.miner_nominal_power_meets_consensus_minimum(
+                        $policy,
+                        ctx.store(),
+                        id,
+                    )?;
+                    if !ok {
+                        return Ok(());
+                    }
+                    let power = claim.quality_adj_power.clone();
+                    let miner_state: miner::State = ctx
+                        .state_manager
+                        .get_actor_state_from_address($ts, &miner.into())
+                        .map_err(map_err)?;
+                    let debt = miner_state.fee_debt();
+                    if !debt.is_zero() {
+                        // fee debt don't add the miner to power table
+                        return Ok(());
+                    }
+                    let miner_info = miner_state.info(ctx.store()).map_err(map_err)?;
+                    // check consensus faults
+                    if $ts.epoch() <= miner_info.consensus_fault_elapsed {
+                        return Ok(());
+                    }
+                    $id_power_worker_mappings.push((id, power, miner_info.worker.into()));
+                    Ok(())
+                })?;
+            };
+        }
+
+        let tsk = f3_tsk.try_into()?;
+        let ts = ctx.chain_index().load_required_tipset(&tsk)?;
+        let state: power::State = ctx.state_manager.get_actor_state(&ts)?;
+        let mut id_power_worker_mappings = vec![];
+        match &state {
```

Comment:
nit: since this match is so long - perhaps it would be more readable to separate this into helper methods?

Reply:
This is tracked by https://github.com/ChainSafe/fil-actor-states/issues/324